### PR TITLE
Introduces PartitionIdSet a special purpose set for partition IDs

### PR DESCRIPTION
**Pull request by vbekiaris**
Backed by `BitSet`, `PartitionIdSet` operations are usually an order of magnitude
faster vs HashSet while being more memory efficient in most cases.
Additionally, primitive `int` versions of `add`, `contains`, `remove` and
`iterator` methods are supplied so clients can avoid autoboxing costs.

### CODE REVIEW

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/util/collection/PartitionIdSetTest.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.PrimitiveIterator;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotSame;
+import static org.junit.Assert.assertTrue;
+
+public class PartitionIdSetTest {
```
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/util/collection/PartitionIdSetTest.java]: Missing `@Category`

**Comment by mmedenjak** [comment]
Needs rebase

File: hazelcast/src/main/java/com/hazelcast/util/collection/ImmutablePartitionIdSet.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ */
+
+package com.hazelcast.util.collection;
+
+import java.util.Collection;
+
+/**
+ * An immutable {@link PartitionIdSet}.
+ */
+public final class ImmutablePartitionIdSet extends PartitionIdSet {
```
[the remark this answers is not in the document]
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/util/collection/ImmutablePartitionIdSet.java]: ... and `removeAll`.

[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/util/collection/ImmutablePartitionIdSet.java]: good catch, thanks!

File: hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java [hunk 3]
```diff
@@ -207,24 +207,24 @@ private void addResultsOfPredicate(List<Future<Result>> futures, Result result,
             if (queryResult == null) {
                 continue;
             }
-            Collection<Integer> queriedPartitionIds = queryResult.getPartitionIds();
+            PartitionIdSet queriedPartitionIds = queryResult.getPartitionIds();
             if (queriedPartitionIds != null) {
-                if (!hasAllBitsSet(finishedPartitionIds, queriedPartitionIds)) {
```
[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java]: nice catch, all of `BitSetUtil` is no longer used, will remove it along with its test.

**Comment by vbekiaris** [comment]
@taburet @mdogan thanks for the reviews!

**Pull request merged**